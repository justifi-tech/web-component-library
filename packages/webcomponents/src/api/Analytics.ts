import { ComponentInterface } from '@stencil/core';
import webcomponentsPackageJson from '../../package.json';
import { AnalyticsService } from './services/analytics.service';

interface iBasicData {
  resource_id: string;
  component_name: string;
  component_version: string;
  client_user_agent: string;
  client_platform: string;
  client_origin: string;
  error?: any;
}

/**
 * @class JustifiAnalytics
 * @description A class to track analytics for a component
 * @param { component } ComponentInterface - The component to track
 *
 */
class JustifiAnalytics {
  componentInstance: any;
  eventEmitters: string[];
  service: AnalyticsService;
  basicData: iBasicData;

  constructor(component: ComponentInterface) {
    this.service = new AnalyticsService();
    this.componentInstance = component;
    this.setUpBasicData();
    this.setupLifecycleTracking();
    this.trackCustomEvents();
  }

  get resourceId() {
    return (
      this.componentInstance.accountId ||
      this.componentInstance.businessId ||
      this.componentInstance.paymentId ||
      this.componentInstance.payoutId ||
      this.componentInstance.checkoutId
    );
  }

  setUpBasicData() {
    this.basicData = {
      component_name: this.componentInstance.tagName,
      component_version: webcomponentsPackageJson.version,
      client_user_agent: navigator.userAgent,
      // navigator.platform is deprecated, use navigator.userAgent instead
      client_platform: navigator.userAgent,
      client_origin: window.location.origin,
      resource_id: this.resourceId,
    };
  }

  handleCustomEvent = async (data: any) => {
    // Track the event
    await this.service.record(data);
  };

  private trackCustomEvents() {
    this.eventEmitters = ['submitted', 'error-event'];

    // for each event, add an event listener
    this.eventEmitters.forEach((eventName) => {
      // if this.componentInstance.addEventListener is a function add the event listener
      if (typeof this.componentInstance.addEventListener === 'function') {
        this.componentInstance.addEventListener(eventName, (event: any) =>
          this.handleCustomEvent({
            event_type: eventName,
            data: { ...this.basicData, error: event.detail },
          })
        );
      }
    });
  }

  setupLifecycleTracking() {
    const originalComponentDidLoad = this.componentInstance.componentDidLoad;

    this.componentInstance.componentDidLoad = () => {
      this.service.record({ event_type: 'init', data: this.basicData });

      if (originalComponentDidLoad) {
        return originalComponentDidLoad.apply(this.componentInstance);
      }
    };
  }

  cleanup() {
    // Remove event listeners
    this.eventEmitters.forEach((event) => {
      this.componentInstance.removeEventListener(event, this.handleCustomEvent);
    });
  }
}

export default JustifiAnalytics;
