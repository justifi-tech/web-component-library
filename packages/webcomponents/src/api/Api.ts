import { v4 as uuidv4 } from 'uuid';
import { PagingInfo } from './Pagination';

export interface IApiResponse<T> {
  data?: T;
  error?: IErrorObject | IServerError;
  page_info?: PagingInfo;
  errors?: string[];
  id?: number;
  type?: string;
}

export type IServerError = string;

export interface IErrorObject {
  message: string;
  code: string;
  param?: string;
}

export interface IApiResponseCollection<T> extends IApiResponse<T> {
  page_info: PagingInfo;
}

const Api = (authToken: string, apiOrigin: string) => {
  async function getAuthorizationHeader() {
    return {
      'Authorization': `Bearer ${authToken}`,
      'Idempotency-Key': uuidv4(),
      'Content-Type': 'application/json',
    };
  }

  async function makeRequest(endpoint: string, method: string, params?: any, body?: any, signal?: AbortSignal) {
    const url = `${apiOrigin}/v1/${endpoint}`;
    const requestUrl = params ? `${url}?${new URLSearchParams(params)}` : url;
    const response = await fetch(requestUrl, {
      method: method,
      headers: await getAuthorizationHeader(),
      body: body,
      signal: signal,
    });

    if (response) {
      return response.status === 204 ? {} : response.json();
    }
    handleError(requestUrl);
  }

  async function get(endpoint: string, params?: any, signal?: AbortSignal) {
    return makeRequest(endpoint, 'GET', params, null, signal);
  }

  async function post(endpoint: string, body?: any, params?: any, signal?: AbortSignal) {
    return makeRequest(endpoint, 'POST', params, body, signal);
  }

  async function patch(endpoint: string, body?: any, params?: any, signal?: AbortSignal) {
    return makeRequest(endpoint, 'PATCH', params, body, signal);
  }

  async function destroy(endpoint: string, params?: any, signal?: AbortSignal) {
    return makeRequest(endpoint, 'DELETE', params, null, signal);
  }

  return { get, post, patch, destroy };
};

function handleError(requestUrl: string): void {
  console.error(`Error fetching from ${requestUrl}`);
}

export default Api;
