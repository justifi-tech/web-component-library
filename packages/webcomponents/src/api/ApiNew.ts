import { configState } from '../components/config-provider/config-state';
import { generateId } from '../utils/utils';
import { PagingInfo } from './Pagination';

export interface IApiResponse<T> {
  data: T;
  error?: IErrorObject | IServerError;
  page_info?: PagingInfo;
  errors?: string[];
  id: number | string;
  type: string;
}

export type IServerError = string;

export interface IErrorObject {
  message: string;
  code: string;
  param?: string;
}

export interface IApiResponseCollection<T> extends IApiResponse<T> {
  page_info: PagingInfo;
}

interface RequestProps {
  endpoint: string;
  method: string;
  params?: any;
  body?: any;
  signal?: AbortSignal;
  headers?: HeadersInit;
}

const Api = (authToken: string) => {
  
  const apiOrigin = configState.apiOrigin;
  
  async function getAuthorizationHeader() {
    if (!authToken) {
      return {
        'Content-Type': 'application/json',
      };
    }

    return {
      Authorization: `Bearer ${authToken}`,
      'Idempotency-Key': generateId(),
      'Content-Type': 'application/json',
    };
  }

  async function makeRequest(props: RequestProps) {
    const { endpoint, method, params, body, signal, headers } = props;

    const url = `${apiOrigin}/v1/${endpoint}`;
    const requestUrl = params ? `${url}?${new URLSearchParams(params)}` : url;

    const defaultHeaders = await getAuthorizationHeader();
    const mergedHeaders = { ...defaultHeaders, ...headers };

    const response = await fetch(requestUrl, {
      method: method,
      headers: mergedHeaders,
      body: body,
      signal: signal,
    });

    if (response) {
      return response.status === 204 ? {} : response.json();
    }
    handleError(requestUrl);
  }

  async function get({
    endpoint,
    params,
    signal,
    headers,
  }: {
    endpoint: string;
    params?: any;
    signal?: AbortSignal;
    headers?: HeadersInit;
  }) {
    return makeRequest({ endpoint, method: 'GET', params, signal, headers });
  }

  async function post({
    endpoint,
    body,
    params,
    signal,
    headers,
  }: {
    endpoint: string;
    body?: any;
    params?: any;
    signal?: AbortSignal;
    headers?: HeadersInit;
  }) {
    return makeRequest({
      endpoint,
      method: 'POST',
      params,
      body: JSON.stringify(body),
      signal,
      headers,
    });
  }

  async function put({
    endpoint,
    body,
    params,
    signal,
  }: {
    endpoint: string;
    body?: any;
    params?: any;
    signal?: AbortSignal;
  }) {
    return makeRequest({
      endpoint,
      method: 'PUT',
      params,
      body: JSON.stringify(body),
      signal,
    });
  }

  async function patch({
    endpoint,
    body,
    params,
    signal,
  }: {
    endpoint: string;
    body?: any;
    params?: any;
    signal?: AbortSignal;
  }) {
    return makeRequest({
      endpoint,
      method: 'PATCH',
      params,
      body: JSON.stringify(body),
      signal,
    });
  }

  async function destroy({
    endpoint,
    params,
    signal,
  }: {
    endpoint: string;
    params?: any;
    signal?: AbortSignal;
  }) {
    return makeRequest({ endpoint, method: 'DELETE', params, signal });
  }

  return { get, post, put, patch, destroy };
};

function handleError(requestUrl: string): void {
  console.error(`Error fetching from ${requestUrl}`);
}

export default Api;
