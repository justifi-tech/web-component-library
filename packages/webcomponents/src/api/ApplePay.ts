import { PagingInfo } from "./Pagination";

export enum ApplePaySessionStatus {
  STATUS_SUCCESS = 'STATUS_SUCCESS',
  STATUS_FAILURE = 'STATUS_FAILURE',
}

export enum ApplePayButtonType {
  PLAIN = 'plain',
  BUY = 'buy',
  SET_UP = 'set-up',
  DONATE = 'donate',
  CHECK_OUT = 'check-out',
  BOOK = 'book',
  SUBSCRIBE = 'subscribe',
}

export enum ApplePayButtonStyle {
  BLACK = 'black',
  WHITE = 'white',
  WHITE_OUTLINE = 'white-outline',
}

export enum ApplePayMerchantCapability {
  SUPPORTS_3DS = 'supports3DS',
  SUPPORTS_EMV = 'supportsEMV',
  SUPPORTS_CREDIT = 'supportsCredit',
  SUPPORTS_DEBIT = 'supportsDebit',
}

export interface IApplePayLineItem {
  label: string;
  amount: string;
  type?: 'final' | 'pending';
}

export interface IApplePayShippingMethod {
  label: string;
  amount: string;
  identifier: string;
  detail?: string;
}

export interface IApplePayPaymentRequest {
  countryCode: string;
  currencyCode: string;
  merchantCapabilities: ApplePayMerchantCapability[];
  supportedNetworks: string[];
  total: IApplePayLineItem;
  lineItems?: IApplePayLineItem[];
  shippingMethods?: IApplePayShippingMethod[];
  applicationData?: string;
  supportedCountries?: string[];
}

export interface IApplePaySession {
  canMakePayments(): boolean;
  canMakePaymentsWithActiveCard(merchantIdentifier: string): Promise<boolean>;
  supportsVersion(version: number): boolean;
  begin(): void;
  abort(): void;
  completeMerchantValidation(merchantSession: IMerchantSession): void;
  completePayment(result: { status: ApplePaySessionStatus }): void;
  completeShippingMethodSelection(update: any): void;
  completePaymentMethodSelection(update: any): void;

  // Event handlers
  onvalidatemerchant?: (event: IApplePayValidateEvent) => void;
  onpaymentmethodselected?: (event: IApplePayMethodSelectedEvent) => void;
  onshippingmethodselected?: () => void;
  onpaymentauthorized?: (event: any) => void;
  oncancel?: (event: IApplePayCancelEvent) => void;
}

export interface IApplePayConfig {
  merchantIdentifier: string;
  displayName: string;
  initiative: string;
  initiativeContext: string;
  buttonType?: ApplePayButtonType;
  buttonStyle?: ApplePayButtonStyle;
  buttonLocale?: string;
}

export interface IApplePayError {
  code: string;
  message: string;
}

export interface IMerchantSession {
  displayName: string;
  domainName: string;
  epochTimestamp: number;
  expiresAt: number;
  merchantIdentifier: string;
  merchantSessionIdentifier: string;
  nonce: string;
  operationalAnalyticsIdentifier: string;
  pspId: string;
  retries: number;
  signature: string;
}

export interface IApplePayTokenData {
  data: string;
  header: {
    publicKeyHash: string;
    ephemeralPublicKey: string;
    transactionId: string;
  };
  signature: string;
  version: string;
}

export interface IApplePayTokenPaymentMethod {
  displayName: string;
  network: string;
  type: string;
}

export interface IApplePayToken {
  paymentData: IApplePayTokenData;
  paymentMethod: IApplePayTokenPaymentMethod;
  transactionIdentifier: string;
}

export interface IApplePayValidateEvent {
  validationURL: string;
}

export interface IApplePayCancelEvent {
  sessionError?: {
    code: 'unknown' | 'invalidMerchantSession' | 'userCancel' | string;
    info?: string;
  };
}

export interface IApplePayMethodSelectedEvent {
  paymentMethod: IApplePayTokenPaymentMethod;
}


export interface IApplePayPaymentProcessRequest {
  paymentData: IApplePayTokenData;
  paymentMethod: IApplePayTokenPaymentMethod;
  transactionIdentifier: string;
  product_details?: {
    name: string;
    price: number;
    description: string;
  };
  description?: string;
}

export interface IApplePayPaymentResponse {
  id: string;
  type: string;
  page_info: PagingInfo | null;
  data: {
    account_id: string;
    token: string;
  }
}

export interface IApplePayService {
  validateMerchant(authToken: string, accountId: string): Promise<IMerchantSession>;

  processPayment(
    authToken: string,
    accountId: string,
    payload: IApplePayPaymentProcessRequest
  ): Promise<{ success: boolean; data: IApplePayPaymentResponse }>;
}

export class ApplePayConfig implements IApplePayConfig {
  public merchantIdentifier: string;
  public displayName: string;
  public initiative: string;
  public initiativeContext: string;
  public buttonType?: ApplePayButtonType;
  public buttonStyle?: ApplePayButtonStyle;
  public buttonLocale?: string;

  constructor(data: IApplePayConfig) {
    this.merchantIdentifier = data.merchantIdentifier;
    this.displayName = data.displayName;
    this.initiative = data.initiative;
    this.initiativeContext = data.initiativeContext;
    this.buttonType = data.buttonType || ApplePayButtonType.PLAIN;
    this.buttonStyle = data.buttonStyle || ApplePayButtonStyle.BLACK;
    this.buttonLocale = data.buttonLocale || 'en-US';
  }

  public get isValid(): boolean {
    return !!(
      this.merchantIdentifier &&
      this.displayName &&
      this.initiative &&
      this.initiativeContext
    );
  }
}

export class ApplePayPaymentRequest implements IApplePayPaymentRequest {
  public countryCode: string;
  public currencyCode: string;
  public merchantCapabilities: ApplePayMerchantCapability[];
  public supportedNetworks: string[];
  public total: IApplePayLineItem;
  public lineItems?: IApplePayLineItem[];
  public shippingMethods?: IApplePayShippingMethod[];
  public applicationData?: string;
  public supportedCountries?: string[];

  constructor(data: IApplePayPaymentRequest) {
    this.countryCode = data.countryCode;
    this.currencyCode = data.currencyCode;
    this.merchantCapabilities = data.merchantCapabilities;
    this.supportedNetworks = data.supportedNetworks;
    this.total = data.total;
    this.lineItems = data.lineItems;
    this.shippingMethods = data.shippingMethods;
    this.applicationData = data.applicationData;
    this.supportedCountries = data.supportedCountries;
  }

  public get isValid(): boolean {
    return !!(
      this.countryCode &&
      this.currencyCode &&
      this.merchantCapabilities?.length &&
      this.supportedNetworks?.length &&
      this.total?.label &&
      this.total?.amount
    );
  }
}

export class ApplePayHelpers {
  static isApplePaySupported(): boolean {
    return (
      typeof window !== 'undefined' &&
      'ApplePaySession' in window &&
      window.ApplePaySession?.supportsVersion(3)
    );
  }

  static canMakePayments(): boolean {
    if (!this.isApplePaySupported()) {
      return false;
    }
    return window.ApplePaySession?.canMakePayments() || false;
  }

  static async canMakePaymentsWithActiveCard(
    merchantIdentifier: string
  ): Promise<boolean> {
    if (!this.isApplePaySupported()) {
      return false;
    }

    try {
      return (
        (await window.ApplePaySession?.canMakePaymentsWithActiveCard(
          merchantIdentifier
        )) || false
      );
    } catch (error) {
      console.error('Error checking Apple Pay active card:', error);
      return false;
    }
  }

  static formatAmount(amount: number): string {
    return (amount / 100).toFixed(2);
  }

  static parseAmount(amount: string): number {
    return Math.round(parseFloat(amount) * 100);
  }

  static createLineItem(
    label: string,
    amount: number,
    type: 'final' | 'pending' = 'final'
  ): IApplePayLineItem {
    return {
      label,
      amount: this.formatAmount(amount),
      type,
    };
  }

  static getDefaultSupportedNetworks(): string[] {
    return ['amex', 'discover', 'masterCard', 'visa'];
  }

  static getDefaultMerchantCapabilities(): ApplePayMerchantCapability[] {
    return [
      ApplePayMerchantCapability.SUPPORTS_3DS,
      ApplePayMerchantCapability.SUPPORTS_EMV,
      ApplePayMerchantCapability.SUPPORTS_CREDIT,
      ApplePayMerchantCapability.SUPPORTS_DEBIT,
    ];
  }
}

declare global {
  interface Window {
    ApplePaySession?: {
      new (
        version: number,
        paymentRequest: IApplePayPaymentRequest
      ): IApplePaySession;
      canMakePayments(): boolean;
      canMakePaymentsWithActiveCard(
        merchantIdentifier: string
      ): Promise<boolean>;
      supportsVersion(version: number): boolean;
      STATUS_SUCCESS: number;
      STATUS_FAILURE: number;
    };
  }
}
