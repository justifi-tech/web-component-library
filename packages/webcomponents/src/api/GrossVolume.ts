export interface GrossVolumeReportDate {
  date: string;
  value: number;
}

export interface GrossVolumeReport {
  total: number;
  dates: GrossVolumeReportDate[];
}
