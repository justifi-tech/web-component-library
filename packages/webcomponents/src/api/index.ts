export * from './Api';
export * from './ComponentEvents';
export * from './ComponentError';
export {
  IBnpl,
  Completion,
  ICheckout,
  Checkout,
  ICheckoutPaymentMode,
  ICheckoutPaymentModeParam,
  ICheckoutStatus,
  CompletionStatuses,
  ICheckoutCompleteResponse,
  ILoadedEventResponse,
  CheckoutsQueryParams,
  ICheckoutPaymentMethod,
} from './Checkout';
export {
  CardBrand as CheckoutCardBrand,
  AccountType as BankAccountType,
} from './Checkout';
export * from './Insurance';
export * from './Pagination';
export * from './Payment';
export * from './PaymentBalanceTransaction';
export * from './Payout';
export * from './PayoutBalanceTransaction';
export * from './Terminal';
export * from './TerminalOrder';
export * from './TerminalModel';
export * from './SubAccount';
export * from './Document';
export * from './Refund';
export * from './Business';
export * from './Identity';
export * from './Dispute';
export * from './GrossVolume';
export * from './ApplePay';
export * from './GooglePay';
// In a future PR I would like to update the naming of the BankAccount Payment Method class in Payment.ts so this import can be updated to be written the same as the others in this file.
export { BankAccount, IBankAccount } from './BankAccount';
