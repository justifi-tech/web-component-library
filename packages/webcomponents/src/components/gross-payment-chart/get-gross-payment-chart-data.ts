import { ComponentErrorSeverity } from '../../api/ComponentError';
import { getErrorCode, getErrorMessage } from '../../api/services/utils';

export const makeGetGrossPaymentChartData =
  ({ id, authToken, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await service.fetchGrossVolumeChartData(id, authToken);

      if (!response.error) {
        onSuccess(response.data);
      } else {
        const errorMessage = getErrorMessage(response.error);
        const code = getErrorCode(response.error?.code);
        onError({
          error: errorMessage,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    } catch (error) {
      const code = getErrorCode(error?.code);
      const errorMessage = getErrorMessage(error);
      onError({
        error: errorMessage,
        code,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
  };
