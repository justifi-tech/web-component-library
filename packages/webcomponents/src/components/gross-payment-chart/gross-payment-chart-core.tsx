import { Component, Host, Prop, State, Watch, h } from '@stencil/core';
import { Chart, BarController, Colors, BarElement, CategoryScale, LinearScale, Legend, Tooltip, Title, ChartConfiguration } from 'chart.js'
import { GrossVolumeReport, GrossVolumeReportDate } from '../../api/GrossVolume';
import { generateChartOptions } from './chart-utils';
import { ErrorState, LoadingState } from '../details/utils';

Chart.register(
  Colors,
  BarController,
  BarElement,
  CategoryScale,
  LinearScale,
  Legend,
  Tooltip,
  Title
)
@Component({
  tag: 'gross-payment-chart-core',
  styleUrl: 'gross-payment-chart.scss',
})

export class GrossPaymentChartCore {
  chart: Chart
  chartRef: HTMLCanvasElement

  @Prop() getGrossPayment: Function;
  @State() total: number;
  @State() dates: GrossVolumeReportDate[];
  @State() endDate: string;
  @State() loading: boolean = true;
  @State() errorMessage: string = '';

  componentDidLoad() {
    if (this.getGrossPayment) {
      this.fetchData();
    }
  }

  @Watch('getGrossPayment')
  propChanged() {
    if (this.getGrossPayment) {
      this.fetchData();
    }
  }

  async fetchData() {
    this.loading = true;

    this.getGrossPayment({
      onSuccess: (data: GrossVolumeReport) => {
        this.total = data.total;
        this.dates = data.dates;
        this.endDate = this.dates[this.dates.length - 1].date;
        this.loading = false;
        this.initChart();
        this.renderChart();
      },
      onError: (error: string) => {
        this.errorMessage = error;
        this.loading = false;
      }
    });
  }

  renderChart() {
    if (!this.chart) {
      this.initChart()
    } else {
      this.chart.update()
    }
  }

  initChart() {
    if (this.chartRef && this.endDate) {
      this.chart = new Chart(
        this.chartRef.getContext("2d"),
        generateChartOptions(this.total, this.dates, this.endDate) as ChartConfiguration
      );
    } else {
      this.errorMessage = 'No data to display';
    }
  }

  render() {
    return (
      <Host>
        {this.loading && LoadingState()}
        {!this.errorMessage ? <canvas id="chart" ref={(elem) => this.chartRef = elem} /> : ErrorState(this.errorMessage)}
      </Host>
    );
  }
} 
