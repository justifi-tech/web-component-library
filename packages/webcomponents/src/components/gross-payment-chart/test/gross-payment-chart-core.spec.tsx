import { newSpecPage } from '@stencil/core/testing';
import { GrossPaymentChartCore } from '../gross-payment-chart-core';
import { GrossVolumeReport } from '../../../api/GrossVolume';
import { IApiResponse } from '../../../api';
import mockSuccessResponse from '../../../api/mockResponses/mockGrossVolumeReportSuccess.json';
import { makeGetGrossPaymentChartData } from '../get-gross-payment-chart-data';

const mockDataResponse = mockSuccessResponse as IApiResponse<GrossVolumeReport>

// To prevent this error https://github.com/reactchartjs/react-chartjs-2/issues/155
jest.mock('chart.js', () => ({
  Chart: () => null,
}))

describe('gross-payment-chart', () => {
  it('should initialize with default states', async () => {
    const page = await newSpecPage({
      components: [GrossPaymentChartCore],
      html: `<gross-payment-chart-core />`,
    });
    expect(page.rootInstance.loading).toBeTruthy();
    expect(page.rootInstance.errorMessage).toBe('');
  });

  it('updates the state properly', async () => {
    const mockGetSuccessResponse = {
      fetchGrossVolumeChartData: jest.fn().mockResolvedValue(mockDataResponse)
    };

    const getGrossPayment = makeGetGrossPaymentChartData({
      id: 'my-account-id',
      authToken: 'my-token',
      service: mockGetSuccessResponse
    });

    const page = await newSpecPage({
      components: [GrossPaymentChartCore],
      html: `<gross-payment-chart-core />`,
    });

    page.rootInstance.componentWillLoad = () => { };
    page.rootInstance.getGrossPayment = getGrossPayment;
    page.rootInstance.fetchData();
    await page.waitForChanges();

    expect(page.rootInstance.total).toEqual(mockDataResponse.data.total);
    expect(page.rootInstance.dates).toEqual(mockDataResponse.data.dates);
    expect(page.rootInstance.endDate).toEqual(mockDataResponse.data.dates[mockDataResponse.data.dates.length - 1].date);
  });

  it('matches the snapshot when in loading state', async () => {
    const page = await newSpecPage({
      components: [GrossPaymentChartCore],
      html: `<gross-payment-chart-core></gross-payment-chart-core>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  // tests error state
  it('matches the snapshot when in error state', async () => {
    const getGrossPayment = makeGetGrossPaymentChartData({
      id: 'my-account-id',
      authToken: 'my-token',
      service: {
        fetchGrossVolumeChartData: jest.fn().mockRejectedValue('Error message')
      }
    });

    const page = await newSpecPage({
      components: [GrossPaymentChartCore],
      html: `<gross-payment-chart-core></gross-payment-chart-core>`,
    });

    page.rootInstance.componentWillLoad = () => { };
    page.rootInstance.getGrossPayment = getGrossPayment;
    page.rootInstance.fetchData();
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });
});
