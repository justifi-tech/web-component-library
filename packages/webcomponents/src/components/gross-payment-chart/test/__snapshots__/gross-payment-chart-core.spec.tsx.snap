// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gross-payment-chart matches the snapshot when in error state 1`] = `
<gross-payment-chart-core>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-empty-state" style="font-size: 1.2rem;">
    <div>
      Error trying to fetch data : Error message
    </div>
  </main>
</gross-payment-chart-core>
`;

exports[`gross-payment-chart matches the snapshot when in loading state 1`] = `
<gross-payment-chart-core>
  <main class="d-flex justify-content-center p-4 text-center" part="detail-loading-state" style="font-size: 1.2rem;">
    <div class="spinner-border spinner-border-sm" part="detail-loading-spinner" role="status">
      <span class="visually-hidden">
        Loading...
      </span>
    </div>
  </main>
  <canvas id="chart"></canvas>
</gross-payment-chart-core>
`;
